src/Symfony/Component/Cache/Adapter/TagAwareAdapter.php

```diff
@@ -363,39 +364,39 @@ private function getTagVersions(array $tagsByKey, bool $persistTags): array
         $tags = [];
         foreach ($tagVersions as $tag => $version) {
             $tags[$tag.static::TAGS_PREFIX] = $tag;
+            if ($fetchTagVersions) {
+                continue;
+            }
             $knownTagVersion = $this->knownTagVersions[$tag] ?? [0, null];
-            if ($fetchTagVersions || $knownTagVersion[1] !== $version || $now - $knownTagVersion[0] >= $this->knownTagVersionsTtl) {
+            if ($knownTagVersion[1] !== $version || $now >= $knownTagVersion[0]) {
                 // reuse previously fetched tag versions up to the ttl
                 $fetchTagVersions = true;
             }
-            unset($this->knownTagVersions[$tag]); // For LRU tracking
-            if ([0, null] !== $knownTagVersion) {
-                $this->knownTagVersions[$tag] = $knownTagVersion;
-            }
         }
 
         if (!$fetchTagVersions) {
             return $tagVersions;
         }
 
+        if (self::MAX_NUMBER_OF_KNOWN_TAG_VERSIONS < \count($this->knownTagVersions)) {
+            $this->knownTagVersions = array_filter($this->knownTagVersions, static fn ($v) => $now < $v[0]);
```

Comment:
Potentially, this will barely remove any items from the cache, which means that we're going to run this loop quite often in these situations. That's a clear path for high CPU usage to me, and why I preferred cutting the array in half: cutting in half means we're going to potentially make the cache less performant, but at least it will never make things worse that without the cache.

Reply:
I can't argue with you - you're absolutely right. The only thing I could propose is to run pruning episodically like many GC implementations do. For me, the main purpose of pruning is prevention of memory leak in long running workers - they just need a chance to free up memory from time to time.